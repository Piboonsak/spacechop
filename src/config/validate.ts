import Config from '../types/Config';

export const validate = (config) => {
  try {
    Config.check(config);
  } catch (err) {
    throw new Error(`${err.key}: ${err.message}`);
  }
};

let valid = true;

export default (config) => {
  try {
    validate(config);
    if (!valid) {
      valid = true;
      console.log('Config is valid again');
    }
    return true;
  } catch (err) {
    valid = false;
    console.log('\nThere is an error in /config.yml:');
    console.log('--------------------------------');
    console.log(err.message);
    console.log('--------------------------------');
  }
  return false;
};
