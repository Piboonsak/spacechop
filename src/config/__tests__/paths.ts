import { validate } from '../validate';

describe('validate paths', () => {
  it('should require paths', () => {
    const config = {};
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should throw on invalid paths array', () => {
    let config = { paths: {}, sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: [], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: [''], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: ['/'], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: ['/:image'], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: [':preset'], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should not throw on valid config', () => {
    const defaultConfig = {
      paths: ['/:preset'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 200,
              height: 200,
            },
          }],
        },
      },
    };
    expect(() => validate({ ...defaultConfig })).not.toThrow();
    expect(() => validate({
      ...defaultConfig,
      paths: ['/:preset/:image'],
    })).not.toThrow();
  });
});
