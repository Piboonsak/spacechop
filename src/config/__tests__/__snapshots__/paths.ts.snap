// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate paths should require paths 1`] = `"paths: Expected array, but was undefined"`;

exports[`validate paths should throw on invalid paths array 1`] = `"paths: Expected array, but was object"`;

exports[`validate paths should throw on invalid paths array 2`] = `"paths: Requires at least one path"`;

exports[`validate paths should throw on invalid paths array 3`] = `"paths.[0]: Cannot be empty"`;

exports[`validate paths should throw on invalid paths array 4`] = `"paths.[0]: Requires :preset in path"`;

exports[`validate paths should throw on invalid paths array 5`] = `"paths.[0]: Requires :preset in path"`;

exports[`validate paths should throw on invalid paths array 6`] = `"paths.[0]: Must start with /"`;
